osu.Server.Spectator/Database/DatabaseAccess.cs

```diff
@@ -154,22 +156,40 @@ public async Task UpdateRoomParticipantsAsync(MultiplayerRoom room)
             }
         }
 
-        public async Task ClearRoomScoresAsync(MultiplayerRoom room)
+        public async Task<multiplayer_playlist_item> GetPlaylistItemFromRoomAsync(long roomId, long playlistItemId)
         {
-            // for now, clear all existing scores out of the playlist item to ensure no duplicates.
-            // eventually we will want to increment to a new playlist item rather than reusing the same one.
-            long playlistItemId = await connection.QuerySingleAsync<long>("SELECT id FROM multiplayer_playlist_items WHERE room_id = @RoomID", new
+            return await connection.QueryFirstOrDefaultAsync<multiplayer_playlist_item>("SELECT * FROM multiplayer_playlist_items WHERE id = @Id AND room_id = @RoomId", new
             {
-                RoomID = room.RoomID,
+                Id = playlistItemId,
+                RoomId = roomId
             });
+        }
+
+        public async Task<long> AddPlaylistItemAsync(multiplayer_playlist_item item)
+        {
+            await connection.ExecuteAsync(
+                "INSERT INTO multiplayer_playlist_items (room_id, beatmap_id, ruleset_id, allowed_mods, required_mods, created_at, updated_at)"
+                + " VALUES (@room_id, @beatmap_id, @ruleset_id, @allowed_mods, @required_mods, NOW(), NOW())",
+                item);
 
-            await connection.ExecuteAsync("DELETE FROM multiplayer_scores WHERE playlist_item_id = @PlaylistItemID", new { PlaylistItemID = playlistItemId });
-            await connection.ExecuteAsync("DELETE FROM multiplayer_scores_high WHERE playlist_item_id = @PlaylistItemID", new { PlaylistItemID = playlistItemId });
+            return await connection.QuerySingleAsync<long>("SELECT max(id) FROM multiplayer_playlist_items WHERE room_id = @room_id", item);
         }
 
-        public Task EndMatchAsync(MultiplayerRoom room)
+        public async Task ExpirePlaylistItemAsync(long playlistItemId)
         {
-            return connection.ExecuteAsync("UPDATE multiplayer_rooms SET ends_at = NOW() WHERE id = @RoomID", new
+            await connection.ExecuteAsync("UPDATE multiplayer_playlist_items SET expired = 1, updated_at = NOW() WHERE id = @PlaylistItemId", new
+            {
+                PlaylistItemId = playlistItemId
+            });
+        }
+
+        public async Task EndMatchAsync(MultiplayerRoom room)
+        {
+            // Remove all non-expired items from the playlist as they have no scores.
+            await connection.ExecuteAsync("DELETE FROM multiplayer_playlist_items WHERE expired = 0");
```

Comment:
This delete query does not filter by room ID, therefore accidentally nuking all non-expired playlist items globally across all rooms.